File: Packages/typescript/siriusmsg-sdk/src/transport.ts
import { createConnection, Socket } from "node:net";
import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";
import { randomUUID } from "node:crypto";
import type { SiriusMsgServiceRequest, SiriusMsgServiceResponse } from "./_models.js";
import { SiriusMsgMalformedFrameError, SiriusMsgTransportError, serviceError } from "./errors.js";
import { validateRequest, validateResponse } from "./validate.js";

export const protocolVersion = 1;
export const defaultRuntimeDir = join(homedir(), "Library", "Application Support", "SiriusMsg");
export const defaultSocketPath = join(defaultRuntimeDir, "siriusmsg.sock");
export const defaultTokenPath = join(defaultRuntimeDir, "service-token.json");

export type Endpoint = { socketPath: string } | { port: number };

export function requestID(prefix = "sdk"): string {
  return `${prefix}-${randomUUID()}`;
}

export function canonicalJSONString(value: unknown): string {
  return JSON.stringify(sortAndDropNullish(value));
}

export function sortAndDropNullish(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortAndDropNullish);
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .filter(([, child]) => child !== undefined && child !== null)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([key, child]) => [key, sortAndDropNullish(child)]),
    );
  }
  return value;
}

export async function loadAuthToken(tokenPath = defaultTokenPath): Promise<string> {
  const parsed = JSON.parse(await readFile(tokenPath, "utf8")) as { token?: unknown };
  if (typeof parsed.token !== "string" || parsed.token.length === 0) {
    throw new SiriusMsgTransportError(`auth token unavailable at ${tokenPath}`);
  }
  return parsed.token;
}

export class NDJSONConnection {
  private socket?: Socket;
  private buffer = "";
  private responses: SiriusMsgServiceResponse[] = [];
  private readers: Array<(value: SiriusMsgServiceResponse) => void> = [];
  private failures: Array<(error: Error) => void> = [];
  private failure?: Error;

  constructor(private readonly endpoint: Endpoint) {}

  async open(): Promise<void> {
    this.socket = "socketPath" in this.endpoint
      ? createConnection(this.endpoint.socketPath)
      : createConnection({ host: "127.0.0.1", port: this.endpoint.port });
    this.socket.setEncoding("utf8");
    this.socket.on("data", (chunk) => this.receive(String(chunk)));
    this.socket.on("error", (error) => this.reject(error));
    this.socket.on("close", () => this.reject(new SiriusMsgTransportError("connection closed")));
    await new Promise<void>((resolve, reject) => {
      this.socket?.once("connect", resolve);
      this.socket?.once("error", reject);
    });
  }

  async authenticate(authToken: string): Promise<void> {
    await this.write({
      protocolVersion,
      requestID: requestID("auth"),
      kind: "authenticate",
      authToken,
    });
    const response = await this.readResponse();
    if (response.kind === "authenticated") {
      return;
    }
    if (response.kind === "error" && response.error) {
      throw serviceError(response.error.code, response.error.message, response.error.diagnosticCode);
    }
    throw new SiriusMsgTransportError("authentication failed");
  }

  async write(frame: SiriusMsgServiceRequest): Promise<void> {
    if (!this.socket) {
      throw new SiriusMsgTransportError("connection is not open");
    }
    validateRequest(frame);
    this.socket.write(`${canonicalJSONString(frame)}\n`);
  }

  readResponse(): Promise<SiriusMsgServiceResponse> {
    const response = this.responses.shift();
    if (response) {
      return Promise.resolve(response);
    }
    if (this.failure) {
      return Promise.reject(this.failure);
    }
    return new Promise((resolve, reject) => {
      this.readers.push(resolve);
      this.failures.push(reject);
    });
  }

  close(): void {
    this.socket?.destroy();
    this.socket = undefined;
  }

  private receive(chunk: string): void {
    this.buffer += chunk;
    while (this.buffer.includes("\n")) {
      const index = this.buffer.indexOf("\n");
      const line = this.buffer.slice(0, index);
      this.buffer = this.buffer.slice(index + 1);
      if (!line) {
        continue;
      }
      try {
        const decoded = JSON.parse(line);
        validateResponse(decoded);
        const reader = this.readers.shift();
        this.failures.shift();
        if (reader) {
          reader(decoded);
        } else {
          this.responses.push(decoded);
        }
      } catch (error) {
        this.reject(error instanceof Error ? error : new SiriusMsgMalformedFrameError(String(error)));
      }
    }
  }

  private reject(error: Error): void {
    this.failure = this.failure ?? error;
    const failures = this.failures.splice(0);
    this.readers.splice(0);
    for (const reject of failures) {
      reject(error);
    }
  }
}

File: Packages/python/siriusmsg-sdk/src/siriusmsg_sdk/_transport.py
"""Hand-written asyncio NDJSON transport for the SiriusMsg local protocol."""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Any, Mapping, Optional
from uuid import uuid4

from pydantic import BaseModel, RootModel, ValidationError

from siriusmsg_sdk._models import SiriusMsgServiceRequest, SiriusMsgServiceResponse
from siriusmsg_sdk.errors import SiriusMsgMalformedFrameError, SiriusMsgTransportError


DEFAULT_RUNTIME_DIR = Path.home() / "Library" / "Application Support" / "SiriusMsg"
DEFAULT_SOCKET_PATH = DEFAULT_RUNTIME_DIR / "siriusmsg.sock"
DEFAULT_TOKEN_PATH = DEFAULT_RUNTIME_DIR / "service-token.json"
PROTOCOL_VERSION = 1


@dataclass(frozen=True)
class Endpoint:
    socket_path: Optional[Path] = None
    port: Optional[int] = None

    @classmethod
    def unix(cls, socket_path: str | Path = DEFAULT_SOCKET_PATH) -> "Endpoint":
        return cls(socket_path=Path(socket_path))

    @classmethod
    def loopback(cls, port: int) -> "Endpoint":
        return cls(port=port)


def request_id(prefix: str = "sdk") -> str:
    return f"{prefix}-{uuid4()}"


def canonical_data(value: Any) -> dict[str, Any]:
    if isinstance(value, BaseModel):
        return _drop_none(value.model_dump(mode="json", exclude_none=True))
    if isinstance(value, RootModel):
        return value.model_dump(mode="json")
    if isinstance(value, Mapping):
        return _drop_none(dict(value))
    raise TypeError(f"unsupported frame type: {type(value).__name__}")


def canonical_json(value: Any) -> str:
    return json.dumps(canonical_data(value), ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def _drop_none(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: _drop_none(child) for key, child in value.items() if child is not None}
    if isinstance(value, list):
        return [_drop_none(child) for child in value]
    if isinstance(value, Enum):
        return value.value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    return value


def load_auth_token(token_path: str | Path = DEFAULT_TOKEN_PATH) -> str:
    try:
        data = json.loads(Path(token_path).read_text())
    except OSError as exc:
        raise SiriusMsgTransportError(f"auth token unavailable at {token_path}") from exc
    token = data.get("token")
    if not isinstance(token, str) or not token:
        raise SiriusMsgTransportError(f"auth token unavailable at {token_path}")
    return token


class NDJSONConnection:
    def __init__(self, endpoint: Endpoint) -> None:
        self.endpoint = endpoint
        self.reader: asyncio.StreamReader | None = None
        self.writer: asyncio.StreamWriter | None = None

    async def __aenter__(self) -> "NDJSONConnection":
        await self.open()
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        await self.close()

    async def open(self) -> None:
        try:
            if self.endpoint.socket_path is not None:
                self.reader, self.writer = await asyncio.open_unix_connection(str(self.endpoint.socket_path))
            elif self.endpoint.port is not None:
                self.reader, self.writer = await asyncio.open_connection("127.0.0.1", self.endpoint.port)
            else:
                raise SiriusMsgTransportError("endpoint missing socket path or port")
        except OSError as exc:
            raise SiriusMsgTransportError(str(exc)) from exc

    async def authenticate(self, auth_token: str) -> None:
        await self.write(
            SiriusMsgServiceRequest(
                protocolVersion=PROTOCOL_VERSION,
                requestID=request_id("auth"),
                kind="authenticate",
                authToken=auth_token,
            )
        )
        response = await self.read_response()
        if response.kind.value != "authenticated":
            if response.kind.value == "error" and response.error is not None:
                from siriusmsg_sdk.errors import error_from_service

                raise error_from_service(
                    response.error.code.value,
                    response.error.message,
                    response.error.diagnosticCode,
                )
            raise SiriusMsgTransportError("authentication failed")

    async def write(self, frame: SiriusMsgServiceRequest | Mapping[str, Any]) -> None:
        if self.writer is None:
            raise SiriusMsgTransportError("connection is not open")
        self.writer.write((canonical_json(frame) + "\n").encode("utf-8"))
        await self.writer.drain()

    async def read_response(self) -> SiriusMsgServiceResponse:
        if self.reader is None:
            raise SiriusMsgTransportError("connection is not open")
        line = await self.reader.readline()
        if not line:
            raise SiriusMsgTransportError("connection closed")
        try:
            decoded = json.loads(line)
            return SiriusMsgServiceResponse.model_validate(decoded)
        except (json.JSONDecodeError, ValidationError) as exc:
            raise SiriusMsgMalformedFrameError(str(exc)) from exc

    async def close(self) -> None:
        if self.writer is not None:
            self.writer.close()
            try:
                await self.writer.wait_closed()
            except OSError:
                pass
        self.reader = None
        self.writer = None
